File: packages/indexer/src/indexer/worker-protocol/fact-groups.ts
import type { IndexPatchFacts } from "../patches";

/**
 * Fact groups in their canonical V3 declaration order.
 *
 * The order is part of the cross-language worker contract. Producers derive
 * presence from the original patch, while consumers use the same list to
 * reject unknown, duplicate, or reordered declarations.
 */
export const projectIndexFactGroups = [
  "prompts",
  "contexts",
  "tools",
  "lint",
  "definitions",
  "relations",
  "sourceRefs",
  "diagnostics",
  "lintFindings",
  "ruleDescriptors",
  "sources",
  "sourceGraph",
] as const satisfies readonly (keyof IndexPatchFacts)[];

/** One fact group declared by a V3 phase summary. */
export type ProjectIndexFactGroup = (typeof projectIndexFactGroups)[number];

/** Mutable reconstruction target for the otherwise readonly patch contract. */
export type MutableIndexPatchFacts = {
  -readonly [TKey in keyof IndexPatchFacts]: IndexPatchFacts[TKey];
};

interface FactEnvelopePresence {
  readonly kind: unknown;
}

interface PhaseSummaryPresence {
  readonly factCount: unknown;
  readonly factGroups?: unknown;
}

const singletonFactGroups = new Set<ProjectIndexFactGroup>([
  "lint",
  "sourceGraph",
]);

/**
 * Returns the defined own fact properties that a new producer must declare.
 *
 * Runtime validation happens before the first event is yielded, preventing a
 * malformed patch from producing a partial transaction.
 */
export function factGroupsFromPatchFacts(
  facts: IndexPatchFacts,
): readonly ProjectIndexFactGroup[] {
  const groups: ProjectIndexFactGroup[] = [];
  for (const group of projectIndexFactGroups) {
    if (!Object.prototype.hasOwnProperty.call(facts, group)) continue;
    const value = facts[group];
    if (value === undefined) continue;
    validateProducerFactGroup(group, value);
    groups.push(group);
  }
  return groups;
}

/**
 * Validates one terminal declaration against all streamed envelopes.
 *
 * Validation completes before reconstruction starts. Declared array groups
 * are initialized as own empty arrays so an explicit clear cannot collapse
 * into legacy omission.
 */
export function prepareFactGroupReconstruction(
  summary: PhaseSummaryPresence,
  envelopes: readonly FactEnvelopePresence[],
): MutableIndexPatchFacts {
  if (
    typeof summary.factCount !== "number" ||
    !Number.isSafeInteger(summary.factCount) ||
    summary.factCount < 0
  ) {
    throw new Error(
      "phase:done summary factCount must be a nonnegative integer",
    );
  }
  if (summary.factCount !== envelopes.length) {
    throw new Error(
      `phase:done summary factCount = ${summary.factCount}, want ${envelopes.length}`,
    );
  }

  const declared = decodeFactGroups(summary.factGroups);
  if (declared === undefined) return {};

  const declaredSet = new Set(declared);
  const counts = new Map<ProjectIndexFactGroup, number>();
  for (const envelope of envelopes) {
    const group = factGroup(envelope.kind);
    if (!group || !declaredSet.has(group)) {
      throw new Error(
        `worker envelope kind ${String(envelope.kind)} is undeclared`,
      );
    }
    counts.set(group, (counts.get(group) ?? 0) + 1);
  }
  for (const group of declared) {
    if (singletonFactGroups.has(group) && counts.get(group) !== 1) {
      throw new Error(
        `worker singleton fact group ${group} must emit exactly one fact`,
      );
    }
  }

  return initializeDeclaredArrays(declared);
}

function decodeFactGroups(
  value: unknown,
): readonly ProjectIndexFactGroup[] | undefined {
  if (value === undefined) return undefined;
  if (!Array.isArray(value)) {
    throw new Error("phase:done summary factGroups must be an array");
  }

  const groups: ProjectIndexFactGroup[] = [];
  let previous = -1;
  for (const candidate of value) {
    const group = factGroup(candidate);
    const position = group ? projectIndexFactGroups.indexOf(group) : -1;
    if (!group || position <= previous) {
      throw new Error(
        "phase:done summary factGroups must use canonical unique groups",
      );
    }
    groups.push(group);
    previous = position;
  }
  return groups;
}

function initializeDeclaredArrays(
  groups: readonly ProjectIndexFactGroup[],
): MutableIndexPatchFacts {
  const facts: MutableIndexPatchFacts = {};
  for (const group of groups) {
    switch (group) {
      case "prompts":
        facts.prompts = [];
        break;
      case "contexts":
        facts.contexts = [];
        break;
      case "tools":
        facts.tools = [];
        break;
      case "definitions":
        facts.definitions = [];
        break;
      case "relations":
        facts.relations = [];
        break;
      case "sourceRefs":
        facts.sourceRefs = [];
        break;
      case "diagnostics":
        facts.diagnostics = [];
        break;
      case "lintFindings":
        facts.lintFindings = [];
        break;
      case "ruleDescriptors":
        facts.ruleDescriptors = [];
        break;
      case "sources":
        facts.sources = [];
        break;
      case "lint":
      case "sourceGraph":
        break;
    }
  }
  return facts;
}

function validateProducerFactGroup(
  group: ProjectIndexFactGroup,
  value: unknown,
): void {
  if (value === null) {
    throw new Error(`patch fact group ${group} cannot be null`);
  }
  const singleton = singletonFactGroups.has(group);
  if (singleton === Array.isArray(value)) {
    const shape = singleton ? "a singleton" : "an array";
    throw new Error(`patch fact group ${group} must be ${shape}`);
  }
}

function factGroup(value: unknown): ProjectIndexFactGroup | undefined {
  switch (value) {
    case "prompts":
    case "contexts":
    case "tools":
    case "lint":
    case "definitions":
    case "relations":
    case "sourceRefs":
    case "diagnostics":
    case "lintFindings":
    case "ruleDescriptors":
    case "sources":
    case "sourceGraph":
      return value;
    default:
      return undefined;
  }
}

File: packages/local/internal/projectindex/eventwire/fact_groups.go
package eventwire

import (
	"bytes"
	"encoding/json"
	"fmt"
	"io"
)

// IndexFactGroup identifies one V3 patch fact field. Values are ordered by the
// cross-language worker contract, not alphabetically.
type IndexFactGroup string

const (
	IndexFactGroupPrompts         IndexFactGroup = "prompts"
	IndexFactGroupContexts        IndexFactGroup = "contexts"
	IndexFactGroupTools           IndexFactGroup = "tools"
	IndexFactGroupLint            IndexFactGroup = "lint"
	IndexFactGroupDefinitions     IndexFactGroup = "definitions"
	IndexFactGroupRelations       IndexFactGroup = "relations"
	IndexFactGroupSourceRefs      IndexFactGroup = "sourceRefs"
	IndexFactGroupDiagnostics     IndexFactGroup = "diagnostics"
	IndexFactGroupLintFindings    IndexFactGroup = "lintFindings"
	IndexFactGroupRuleDescriptors IndexFactGroup = "ruleDescriptors"
	IndexFactGroupSources         IndexFactGroup = "sources"
	IndexFactGroupSourceGraph     IndexFactGroup = "sourceGraph"
)

var canonicalIndexFactGroups = []IndexFactGroup{
	IndexFactGroupPrompts,
	IndexFactGroupContexts,
	IndexFactGroupTools,
	IndexFactGroupLint,
	IndexFactGroupDefinitions,
	IndexFactGroupRelations,
	IndexFactGroupSourceRefs,
	IndexFactGroupDiagnostics,
	IndexFactGroupLintFindings,
	IndexFactGroupRuleDescriptors,
	IndexFactGroupSources,
	IndexFactGroupSourceGraph,
}

// DecodedFactGroups preserves whether a V3 phase summary declared presence.
// Present empty and legacy omission intentionally have different values.
type DecodedFactGroups struct {
	Present bool
	Value   []IndexFactGroup
}

// DecodeFactGroups validates the optional V3 fact-group declaration.
//
// Nil or zero-length bytes identify a legacy omission. Every present value
// must be one canonical, unique JSON string array; null and trailing content
// are rejected so explicit empty presence cannot collapse into omission.
func DecodeFactGroups(raw json.RawMessage) (DecodedFactGroups, error) {
	if len(raw) == 0 {
		return DecodedFactGroups{}, nil
	}
	trimmed := bytes.TrimSpace(raw)
	if bytes.Equal(trimmed, []byte("null")) {
		return DecodedFactGroups{}, fmt.Errorf("factGroups cannot be null")
	}

	var elements []json.RawMessage
	decoder := json.NewDecoder(bytes.NewReader(raw))
	if err := decoder.Decode(&elements); err != nil {
		return DecodedFactGroups{}, fmt.Errorf("decode factGroups: %w", err)
	}
	if err := decoder.Decode(&struct{}{}); err != io.EOF {
		return DecodedFactGroups{}, fmt.Errorf("decode factGroups: expected one JSON value")
	}
	if elements == nil {
		return DecodedFactGroups{}, fmt.Errorf("factGroups must be an array")
	}

	groups := make([]IndexFactGroup, 0, len(elements))
	previous := -1
	for _, element := range elements {
		value, err := decodeFactGroup(element)
		if err != nil {
			return DecodedFactGroups{}, err
		}
		position := indexFactGroupPosition(value)
		if position < 0 {
			return DecodedFactGroups{}, fmt.Errorf("unknown factGroup %q", value)
		}
		if position <= previous {
			return DecodedFactGroups{}, fmt.Errorf("factGroups must be unique and canonically ordered")
		}
		groups = append(groups, value)
		previous = position
	}
	return DecodedFactGroups{Present: true, Value: groups}, nil
}

func decodeFactGroup(raw json.RawMessage) (IndexFactGroup, error) {
	trimmed := bytes.TrimSpace(raw)
	if len(trimmed) == 0 || trimmed[0] != '"' {
		return "", fmt.Errorf("factGroup must be a string")
	}
	var value string
	if err := json.Unmarshal(trimmed, &value); err != nil {
		return "", fmt.Errorf("decode factGroup: %w", err)
	}
	return IndexFactGroup(value), nil
}

func indexFactGroupPosition(group IndexFactGroup) int {
	for index, candidate := range canonicalIndexFactGroups {
		if group == candidate {
			return index
		}
	}
	return -1
}

func reconstructDeclaredFactGroups(
	facts IndexPatchFacts,
	decoded DecodedFactGroups,
	envelopes []IndexFactEnvelope,
) (IndexPatchFacts, error) {
	if !decoded.Present {
		return facts, nil
	}

	declared := make(map[IndexFactGroup]bool, len(decoded.Value))
	counts := make(map[IndexFactGroup]int, len(decoded.Value))
	for _, group := range decoded.Value {
		declared[group] = true
	}
	for _, envelope := range envelopes {
		group := IndexFactGroup(envelope.Kind)
		if indexFactGroupPosition(group) < 0 || !declared[group] {
			return IndexPatchFacts{}, fmt.Errorf("fact envelope group %q is undeclared", envelope.Kind)
		}
		counts[group]++
	}
	for _, group := range decoded.Value {
		if isSingletonFactGroup(group) && counts[group] != 1 {
			return IndexPatchFacts{}, fmt.Errorf(
				"singleton fact group %q emitted %d facts, want exactly one",
				group,
				counts[group],
			)
		}
		initializeDeclaredFactGroup(&facts, group)
	}
	return facts, nil
}

func initializeDeclaredFactGroup(facts *IndexPatchFacts, group IndexFactGroup) {
	switch group {
	case IndexFactGroupPrompts:
		facts.Prompts = nonNilFactSlice(facts.Prompts)
	case IndexFactGroupContexts:
		facts.Contexts = nonNilFactSlice(facts.Contexts)
	case IndexFactGroupTools:
		facts.Tools = nonNilFactSlice(facts.Tools)
	case IndexFactGroupDefinitions:
		facts.Definitions = nonNilFactSlice(facts.Definitions)
	case IndexFactGroupRelations:
		facts.Relations = nonNilFactSlice(facts.Relations)
	case IndexFactGroupSourceRefs:
		facts.SourceRefs = nonNilFactSlice(facts.SourceRefs)
	case IndexFactGroupDiagnostics:
		facts.Diagnostics = nonNilFactSlice(facts.Diagnostics)
	case IndexFactGroupLintFindings:
		facts.LintFindings = nonNilFactSlice(facts.LintFindings)
	case IndexFactGroupRuleDescriptors:
		facts.RuleDescriptors = nonNilFactSlice(facts.RuleDescriptors)
	case IndexFactGroupSources:
		facts.Sources = nonNilFactSlice(facts.Sources)
	case IndexFactGroupLint, IndexFactGroupSourceGraph:
	}
}

func isSingletonFactGroup(group IndexFactGroup) bool {
	return group == IndexFactGroupLint || group == IndexFactGroupSourceGraph
}

func nonNilFactSlice[T any](values []T) []T {
	if values == nil {
		return make([]T, 0)
	}
	return values
}
